### PR TITLE
Add a way to bind callbacks to IQ responses

### PR DESCRIPTION
**Pull request by powdahound**
Useful for fetching the roster, rooms list, and many other things moving forward.

### CODE REVIEW

**Comment by cjoudrey** [comment]
This is interesting.

Is this something specific to HipChat?

File: lib/bot.js [hunk 1]
```diff
@@ -100,6 +100,13 @@ module.exports.Bot = (function() {
       var jid = attrFrom.substring(0, offset);
 
       this.emit('privateMessage', jid, body.getText());
+    } else if (stanza.is('iq')) {
+      // call stored callback function for IQ response
+      // TODO: emit some sort of error if the ID is unexpected?
+      var id = stanza.attrs.id;
+      if (id in this.iq_callbacks) {
+        this.iq_callbacks[id](stanza);
```
Comment (cjoudrey) [inline lib/bot.js]: Could we clean up the callback once we received the reply?

i.e.

``` json
delete this.iq_callbacks[id];
```

Comment (powdahound) [inline lib/bot.js]: Good catch.

**Comment by powdahound** [comment]
None of this is HipChat-specific btw. [IQs](http://xmpp.org/rfcs/rfc6120.html#stanzas-semantics-iq) are an important part of XMPP and are usually used when requesting information and expecting a response; fetching the roster ([example](http://xmpp.org/rfcs/rfc6121.html#roster-syntax-actions-get)), discovering rooms, getting profile info, etc.

**Comment by cjoudrey** [comment]
Really interesting. I actually didn't look into IQs yet.

After looking into the code further, I think we could use the Bot's eventemitter to handle these events instead of creating a new hash.

i.e.

``` javascript
this.once('iq:' + id, callback);
```

and

``` javascript
this.emit('iq:' + id, stanza);
```

This will simplify your code a lot and we won't need to worry about cleaning up the callbacks once the IQ response comes in.

**Comment by cjoudrey** [comment]
Later on it would be interesting to abstract these different IQs messages using the `sendIq` function.

Something like:

``` javascript
Bot.getRoster(channelId, function(err, roster) {

});
```

What do you think?

**Comment by powdahound** [comment]
Cool - wasn't aware of that mechanism but it looks nice (and more inline with the existing evented style). Will check it out once I get to the office.

And yeah, having higher-level functions to request common things is the eventual goal. Bot.getRoster(cb), Bot.getRooms(cb), etc. I'm also adding a method to our server so Wobot can fetch the room nickname to use instead of requiring the user to configure it to match what's on the bot's HipChat account.

**Comment by powdahound** [comment]
Planning to make the other changes taking advantage of the IQ support in other pull requests btw. Feel free to merge if this all works for you, and thanks for the quick feedback!

**Comment by cjoudrey** [comment]
Thanks for contributing! :)

One small thing, most callbacks in node have the format: `(err, response)`.

Just for predictability reasons, could you have `getRoster` callback use `(err, result)`?

Even better, maybe we could parse the stanza for the user to make it easier to use.

i.e.

```
Bot.getRoster('channel', function (err, roster, stanza) {

});
```

In most cases, the user will only need the `name` and `jid` of the users. The `stanza` would still be available as 3rd argument for advanced uses.

The `roster` could be an array of objects:

``` json
[
  {name: '...', jid: '...'},
  {name: '...', jid: '...'},
  {name: '...', jid: '...'}
]
```

What do you think?

**Comment by powdahound** [comment]
Using `(err, result)` works - I suppose it should actually be checking for an error response anyway. Not that our servers ever have trouble. :)

Passing a parsed list of items back as well as the stanza is a good idea. I was going to pass back names/jids but didn't want to hide other data from those who wanted it. Your solution will work well.

**Comment by powdahound** [comment]
OK check that out. Easy to handle error responses to any IQ request and function signatures contain an `err` as expected.

**Comment by cjoudrey** [comment]
Looks awesome. :)

**Comment by powdahound** [comment]
Sweet - starting getting a better feel for common node conventions. If you can merge this and #14 soon I'll get to work on the next batch of things. :)

**Comment by cjoudrey** [comment]
Nice :D

Sure thing! I will merge it tonight.

File: examples/example1.js [hunk 2]
```diff
@@ -1,5 +1,6 @@
 var Bot = require('../lib/').Bot;
 var http = require('http');
+var xmpp = require('node-xmpp');
```
Comment (cjoudrey) [inline examples/example1.js]: We can remove this line now that we have getRoster :)

**Comment by cjoudrey** [comment]
Is there any way of knowing the status of the users when calling `getRoster`?

We could return:

``` json
{
  name: '...',
  jid: '...',
  status: '...'
}
```

It currently isn't sent back in the iq reply. I'm guessing this would be against the RFC...

**Comment by powdahound** [comment]
Nope - can't get status from that call. Status is sent back for every user automatically (`<presence>` stanzas from each user) once the bot sets its initial presence but that data isn't currently being captured. Handling stuff like that wouldn't really be in the scope of this change.

**Comment by cjoudrey** [comment]
Makes sense, we can open another issue for this when the time comes.

**Pull request closed without merging**